tools/kwiver_tool_runner.cxx

```diff
@@ -186,21 +333,59 @@ int main(int argc, char *argv[])
 
   // initialize the global context
   tool_context->m_wtb.set_indent_string( "      " );
-
+  cmd_root = build_command_tree( vpm );
   command_line_parser options( argc, &argv );
 
-  if ( (options.m_applet_name == "help") || (argc == 1) )
+  // If there is only a 'help' as an argument
+  if ( (argc == 1) || (options.m_applet_name[0] == "help") )
   {
-    help_applet( options, tool_context, vpm );
+    tool_runner_usage( tool_context, vpm );
     return 0;
   } // end help code
 
   // ----------------------------------------------------------------------------
   try
   {
     // Create applet based on the name provided
-    applet_factory app_fact;
-    kwiver::tools::kwiver_applet_sptr applet( app_fact.create( options.m_applet_name ) );
+    auto cn = cmd_root;
+    size_t ci;
+    for ( ci = 0; ci < options.m_applet_name.size(); ci++)
+    {
+      auto n = cn->find( options.m_applet_name[ci] );
+      if ( ! n )
+      {
+        break;
+      }
+      cn = n;
+    } // end for
+
+    // If there are no matches or we run out of command words
+    // before we reach a terminal command node.
+    if ( ci == 0 && !cn->is_command() )
+    {
+      // error - no command match at all. Give full help
+      std::cerr << "Command not found.\n\n";
+      tool_runner_usage( tool_context, vpm );
+      exit(-1);
+    }
+    else if ( ci > 0 && !cn->is_command() )
+    {
+      // Partial command match - just subtree help
+      std::cerr << "Command not found. Related commands are as follows:\n";
+      cn->print_subtree_help();
+      exit(-1);
+    }
+
+    // Insert name words[ci+1] [size-1] at start of options.m_applet_args
+    for( size_t ai = ci; ai < options.m_applet_name.size(); ai++)
+    {
+      options.m_applet_args.insert(
+        options.m_applet_args.begin()+1, // skip first "kwiver" element
```

Comment:
What is this first "kwiver" element?  Does this mean the arguments to the function are {"kwiver", "arg1", "arg2",...}?  Is "kwiver" in this list but the name of the applet not in this list?

Reply:
The "kwiver" element is inserted to comply with standard argv semantics (the name of "program") so that command line parsing has something there. If I put an applet name there which has a multi-word name, the command line parser will break each word int a separate argument. The applet name is available to the applet via the applet_name() call.